chunky/src/java/se/llbit/chunky/world/ChunkSelectionTracker.java

```diff
@@ -32,10 +32,100 @@
  */
 public class ChunkSelectionTracker implements ChunkDeletionListener {
 
-  private final Set<ChunkPosition> selected = new HashSet<>();
+  private final ConcurrentHashMap<Long, BitSet> selectedChunksByRegion = new ConcurrentHashMap<>();
   private final Collection<ChunkUpdateListener> chunkUpdateListeners = new LinkedList<>();
   private final Collection<ChunkSelectionListener> selectionListeners = new LinkedList<>();
 
+  /**
+   * @return Whether the selection changed
+   */
+  private boolean setChunk(ChunkPosition pos, boolean selected) {
+    BitSet selectedChunksForRegion = selectedChunksByRegion.computeIfAbsent(ChunkPosition.positionToLong(pos.x >> 5, pos.z >> 5), p -> new BitSet(MCRegion.CHUNKS_X * MCRegion.CHUNKS_Z));
+    int bitIndex = (pos.x & 31) + ((pos.z & 31) << 5);
+    boolean previousValue = selectedChunksForRegion.get(bitIndex);
+    if(previousValue != selected) {
+      selectedChunksForRegion.set(bitIndex, selected);
+
+      if (selectedChunksForRegion.nextSetBit(0) == -1) {
+        //all bits are 0, we don't need to track this region anymore
+        selectedChunksByRegion.remove(ChunkPosition.positionToLong(pos.x >> 5, pos.z >> 5));
+      }
+
+      notifyChunkUpdated(pos);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @return Whether the selection changed
+   */
+  private boolean setChunksWithinRegion(ChunkPosition regionPos, int minX, int maxX, int minZ, int maxZ, boolean selected) {
+    BitSet selectedChunksForRegion = selectedChunksByRegion.computeIfAbsent(regionPos.getLong(), p -> new BitSet(MCRegion.CHUNKS_X * MCRegion.CHUNKS_Z));
+
+    Collection<ChunkPosition> changedChunks = new ArrayList<>();
+    boolean selectionChanged = false;
+
+    for (int chunkX = minX; chunkX < maxX; chunkX++) {
+      for (int chunkZ = minZ; chunkZ < maxZ; chunkZ++) {
+        int bitIndex = (chunkX & 31) + ((chunkZ & 31) << 5);
+        boolean previousValue = selectedChunksForRegion.get(bitIndex);
+        if(previousValue != selected) {
+          selectionChanged = true;
+          selectedChunksForRegion.set(bitIndex, selected);
+
+          changedChunks.add(ChunkPosition.get(chunkX, chunkZ));
+        }
+      }
+    }
+
+    if (selectedChunksForRegion.nextSetBit(0) == -1) {
+      //all bits are 0, we don't need to track this region anymore
+      selectedChunksByRegion.remove(regionPos.getLong());
+    }
+
+    if(selectionChanged) {
+      notifyRegionChunksUpdated(regionPos, changedChunks);
+//      notifyChunksUpdated(changedChunks);
+//      notifyRegionChunksUpdated(regionPos);
```

Comment:
Remove this then?

Reply:
oh - good point, this was some debugging stuff